Make an asynchronous version of _WKInputDelegate._webView:focusRequiresStrongPasswordAssistance:

achristensen07 opened the pull request:
#### 5ceeaba755e447f42f3cdd36cf7b600dad6b9798
<pre>
Make an asynchronous version of _WKInputDelegate._webView:focusRequiresStrongPasswordAssistance:
<a href="https://bugs.webkit.org/show_bug.cgi?id=286494">https://bugs.webkit.org/show_bug.cgi?id=286494</a>
<a href="https://rdar.apple.com/143579852">rdar://143579852</a>

Reviewed by Wenson Hsieh.

With minimal reordering of when _focusedElementInformation is set, we can call
_continueElementDidFocus synchronously or asynchronously and there should be no
behavior change.  To prevent _continueElementDidFocus from being called after a
blur, I add a check to _focusedElementInformation.identifier after the asynchronous
delegate call to do nothing if we have changed focus during the delegate call.
I changed the SetForScope to a CompletionHandlerCallingScope so I can pass it through
an asynchronous call.

* Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _continueElementDidFocus:focusedElementInfo:requiresKeyboard:activityStateChanges:]):

Canonical link: <a href="https://commits.webkit.org/289895@main">https://commits.webkit.org/289895@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5d15f76cd9d053ce76cc0bc49f9993b024d8017

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86724 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6231 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41059 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91569 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37456 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88773 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6499 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14289 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67058 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24828 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89727 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78512 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47371 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4736 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32851 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36574 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75251 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33736 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93464 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13877 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10061 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75848 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14078 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74342 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75043 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19350 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17756 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6662 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13798 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13900 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19160 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13638 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17083 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15423 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -8084,8 +8084,17 @@ - (void)stopDeferringInputViewUpdatesForAllSources
+            strongSelf->_isFocusingElementWithKeyboard = valueToRestore;
+    });
+    _isFocusingElementWithKeyboard = [self _shouldShowKeyboardForElement:information];
 
```
I _suspect_ we'd probably want to move `runModalJavaScriptDialogCallbackIfNeeded` and `stopDeferringInputViewUpdatesScope` into the calling scope as well — otherwise, it might bring back various deadlocks and/or crashes related to showing synchronous JS prompts during element focus.

achristensen07 replied:
And I can do it all with one CompletionHandlerCallingScope!

whsieh reviewed (approved):
This seems reasonable to me, although we probably should:
- Write a basic API or layout test, to exercise the case where an input element is focused and immediately blurred and the WK client implements the new async delegate method. The UI process should not end up with an active input session.
- It would probably be good to run `editing/` and `fast/forms/` with an implementation of this delegate method that waits for at least 1 turn of the main runloop, to see if there's any other edge cases that we've missed.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -8186,10 +8195,44 @@ - (void)_elementDidFocus:(const WebKit::FocusedElementInformation&)information u
+                return;
+            if (focusedElementInformationIdentifier != strongSelf->_focusedElementInformation.identifier)
+                return;
+            [strongSelf _continueElementDidFocus:result focusedElementInfo:focusedElementInfo requiresKeyboard:requiresKeyboard activityStateChanges:activityStateChanges isChangingFocusForScope:WTFMove(isChangingFocusForScope) isFocusingElementWithKeyboardForScope:WTFMove(isFocusingElementWithKeyboardForScope)];
```
This would be more readable with some newlines.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -8207,9 +8250,7 @@ - (void)_elementDidFocus:(const WebKit::FocusedElementInformation&)information u
 
-    BOOL requiresKeyboard = mayContainSelectableText(information.elementType);
     BOOL editableChanged = [self setIsEditable:requiresKeyboard];
-    _focusedElementInformation = information;
```
Hopefully there are no observable side effects from setting `_focusedElementInformation` before calling `willStartInputSession`...

whsieh replied:
Oh, this comment made me realize — we might get stuck in a state where `_isEditable` is set indefinitely, if the element is blurred after focusing but before the input delegate calls the completion handler. We should probably have logic to either revert this to `NO` or move the call to `[self setIsEditable:requiresKeyboard]` into the continuation.

achristensen07 replied:
I'm going to undo this change.

achristensen07 committed:
Make an asynchronous version of _WKInputDelegate._webView:focusRequiresStrongPasswordAssistance:
https://bugs.webkit.org/show_bug.cgi?id=286494
rdar://143579852

Reviewed by Wenson Hsieh.

With minimal reordering of when _focusedElementInformation is set, we can call
_continueElementDidFocus synchronously or asynchronously and there should be no
behavior change.  To prevent _continueElementDidFocus from being called after a
blur, I add a check to _focusedElementInformation.identifier after the asynchronous
delegate call to do nothing if we have changed focus during the delegate call.
I changed the SetForScope to a CompletionHandlerCallingScope so I can pass it through
an asynchronous call.

* Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _continueElementDidFocus:focusedElementInfo:requiresKeyboard:activityStateChanges:]):

Canonical link: https://commits.webkit.org/289895@main
